docs/userguide/mpi_apps.rst

```diff
@@ -1,335 +1,132 @@
+        from subprocess import run
+        with open('stdout.lmp', 'w') as fp, open('stderr.lmp', 'w') as fe:
+            proc = run(['$PARSL_MPI_PREFIX', '-i', 'in.lmp'], stdout=fp, stderr=fe)
+            return proc.returncode
 
-.. code-block:: python
 
-    @parsl.bash_app
-    def cosmic_tagger(workdir: str,
-                      datatype: str = "float32",
-                      batchsize: int = 8,
-                      framework: str = "torch",
-                      iterations: int = 500,
-                      trial: int = 2,
-                      stdout=parsl.AUTO_LOGNAME,
-                      stderr=parsl.AUTO_LOGNAME,
-                      parsl_resource_specification:Dict={}):
-        NRANKS = parsl_resource_specification['num_ranks']
-
-        return f"""
-        module purge
-        module use /soft/modulefiles/
-        module load conda/2023-10-04
-        conda activate
-
-        echo "PARSL_MPI_PREFIX : $PARSL_MPI_PREFIX"
-
-        $PARSL_MPI_PREFIX --cpu-bind numa \
-            python {workdir}/bin/exec.py --config-name a21 \
-                run.id=run_plrs_ParslDemo_g${NRANKS}_{datatype}_b{batchsize}_{framework}_i{iterations}_T{trial} \
-                run.compute_mode=GPU \
-                run.distributed=True \
-                framework={framework} \
-                run.minibatch_size={batchsize} \
-                run.precision={datatype} \
-                mode.optimizer.loss_balance_scheme=light \
-                run.iterations={iterations}
-        """
-
-In this example, we run a simple test that does an exploration over the ``batchsize`` parameter
-while launching the application over 2-4 nodes.
+Run either App by calling with its arguments and a resource specification which defines how to execute it
 
 .. code-block:: python
 
-    def run_cosmic_tagger():
-        futures = {}
-        for num_nodes in [2, 4]:
-            for batchsize in [2, 4, 8]:
-
-                parsl_res_spec = {"num_nodes": num_nodes,
-                                  "num_tasks": num_nodes * 4,
-                                  "ranks_per_node": 4}
-                future = cosmic_tagger(workdir="/home/yadunand/CosmicTagger",
-                                       datatype="float32",
-                                       batchsize=str(batchsize),
-                                       parsl_resource_specification=parsl_res_spec)
-
-
-                print(f"Stdout : {future.stdout}")
-                print(f"Stderr : {future.stderr}")
-                futures[(num_nodes, batchsize)] = future
+    # Resources in terms of nodes and how ranks are to be distributed are set on a per app
+    # basis via the resource_spec dictionary.
+    resource_spec = {
+        "num_nodes" = 2,
+        "ranks_per_node" = 2,
+        "num_ranks" = 4,
+    }
+    future = lammps_mpi_application(File('in.file'), parsl_resource_specification=resource_spec)
 
+Advanced: More Environment Variables
```

Comment:
Very selfish question:

I am usually running VASP (via ASE) on the GPU nodes of Perlmutter. I assume one is still using a `SimpleLauncher()` here since it's an "internal" call to `srun`?

For context, a Slurm script usually looks something like the following:

```python
#!/bin/bash
#SBATCH -N 1
#SBATCH -C gpu
#SBATCH -G 4
#SBATCH -q debug
#SBATCH -J vasp_job
#SBATCH -t 00:30:00
#SBATCH -A matgen_g

module load vasp/6.4.1-gpu
export OMP_NUM_THREADS=8
export OMP_PLACES=threads
export OMP_PROC_BIND=spread
--gpu-bind=single:1
export ASE_VASP_COMMAND="srun -n 4 -c 32 --cpu_bind=cores -G 4 --gpu-bind=none vasp_std"

python run_my_jobs.py
```

where `run_my_jobs.py` is some arbitrary program that in it launches a bunch of `@python_app`s representing one VASP task per App.

Reply:
I think this application is possible by replacing the parts of the VASP command definition which deal with the total ranks to `\$PARSL_MPI_PREFIX`. Escaping the `$` will (hopefully!) result in the environment variable not being resolved with `ASE_VASP_COMMAND` is defined but instead when the contents `ASE_VASP_COMMAND` is used to start a subprocess somewhere deep in the bowels of ASE.

```bash
export ASE_VASP_COMMAND="\$PARSL_MPI_PREFIX -c 32 --cpu_bind=cores -G 4 --gpu-bind=none vasp_std"
```

That said, I'm not sure if `-G` is the "total" number of GPUs or the number per node. You might to switch to `--gpus-per-node` instead of `-G`.